source/extensions/transformers/aws_lambda/api_gateway_transformer.cc

```diff
@@ -0,0 +1,202 @@
+#include "source/extensions/transformers/aws_lambda/api_gateway_transformer.h"
+
+#include "source/common/http/headers.h"
+#include "source/common/http/utility.h"
+#include "source/common/common/base64.h"
+
+#include "source/common/http/header_map_impl.h"
+
+#include "nlohmann/json.hpp"
+using json = nlohmann::json;
+
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace AwsLambda {
+
+HttpFilters::Transformation::TransformerConstSharedPtr
+ApiGatewayTransformerFactory::createTransformer(
+    const Protobuf::Message &config,
+    Server::Configuration::CommonFactoryContext &context) {
+    MessageUtil::downcastAndValidate<const ApiGatewayTransformerProto &>(
+          config, context.messageValidationContext().staticValidationVisitor());
+  return std::make_shared<ApiGatewayTransformer>();
+}
+
+ApiGatewayTransformer::ApiGatewayTransformer(){}
+const Envoy::Http::LowerCaseString ApiGatewayTransformer::AMAZON_ERRORTYPE_HEADER = Envoy::Http::LowerCaseString("x-amzn-errortype");
+
+void ApiGatewayTransformer::format_error(
+  Http::ResponseHeaderMap &response_headers,
+  Buffer::Instance &body,
+  ApiGatewayError &error,
+  Http::StreamFilterCallbacks &stream_filter_callbacks) const {
+    ENVOY_STREAM_LOG(debug, "Returning error with message: " + std::string(error.message), stream_filter_callbacks);
+
+    // clear existing response headers
+    response_headers.clear();
+
+    response_headers.setStatus(error.status_code);
+    response_headers.setContentType("text/plain");
+
+    response_headers.setReferenceKey(AMAZON_ERRORTYPE_HEADER, error.code);
+    body.drain(body.length());
+    auto error_message = absl::StrCat(error.code, ": ", error.message);
+    body.add(error_message);
+    response_headers.setContentLength(body.length());
+  }
+
+void ApiGatewayTransformer::transform(
+    Http::RequestOrResponseHeaderMap &header_map,
+    Http::RequestHeaderMap *request_headers,
+    Buffer::Instance &body,
+    Http::StreamFilterCallbacks &stream_filter_callbacks) const {
+      ENVOY_STREAM_LOG(debug, "Transforming response", stream_filter_callbacks);
+      // check if request header map equals response header map
+      if (*request_headers == header_map) {
```

Comment:
is this just checking for pointer equality? shouldn't we be checking if the header sets are the same? or perhaps.. just don't even call this function on request path (envoy hooks are per request/response)

Reply:
I believe it is using pointer equality to determine if it's being called from the request or response. We only call [this](https://github.com/solo-io/envoy-gloo/blob/9fcbba0f96213546b91a97bcf99fd36009c31bb4/source/extensions/filters/http/aws_lambda/aws_lambda_filter.cc#L180) in the `encodeData` and `encodeTrailers` functions, so it is codified to not be on the request path. The only way I could see this being an issue is if the control plane introduced a bug to specify this transformer as the request transformer. In light of that, I think the defensiveness is warranted

Reply:
I think we can make a new class `ResponseTransformer` (if does not already exist) to extend `Transformer` so we don't need this check and future devs don't need to reason about this potential misuse

Reply:
This refactor is being worked on in a [separate PR](https://github.com/solo-io/envoy-gloo/pull/214)